drivers/char/tpm/tpm_ftpm_optee.c

```diff
@@ -76,37 +81,35 @@ static int ftpm_optee_tpm_op_send(struct tpm_chip *chip, u8 *buf, size_t len)
 		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_NONE }
 	};
 	struct tee_ioctl_invoke_arg submit_command_args;
-	struct tee_shm *shm = NULL;
 
-	mutex_lock(&tpm_tee_context.response_buffer_lock);
+	if (tpm_tee_context.response_length != 0) {
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	if (len > FTPM_SHM_SIZE - FTPM_RESPONSE_SIZE_MAX) {
+		rc = -EIO;
+		goto cleanup;
+	}
 
+	tpm_tee_context.response_offset = len;
 	tpm_tee_context.response_length = 0;
+	memset(tpm_tee_context.shm->kaddr, 0, FTPM_SHM_SIZE);
 
 	submit_command_args.func = FTPM_OPTEE_TA_SUBMIT_COMMAND;
 	submit_command_args.session = tpm_tee_context.optee_session_id;
 	submit_command_args.num_params = TEEC_CONFIG_PAYLOAD_REF_COUNT;
 
-	shm = tee_shm_alloc(
-		tpm_tee_context.optee_context, 
-		len + sizeof(tpm_tee_context.response_buffer), 
-		TEE_SHM_MAPPED | TEE_SHM_DMA_BUF);
-
-	if (IS_ERR(shm)) {
-		rc = -ENOMEM;
-		shm = NULL;
-		goto cleanup;
-	}
-
 	// request
 	submit_command_params[0].u.memref.size = len;
-	submit_command_params[0].u.memref.shm = shm;
+	submit_command_params[0].u.memref.shm = tpm_tee_context.shm;
 	submit_command_params[0].u.memref.shm_offs = 0;
 	memcpy(submit_command_params[0].u.memref.shm->kaddr, buf, len);
 
 	// response
-	submit_command_params[1].u.memref.size = sizeof(tpm_tee_context.response_buffer);
-	submit_command_params[1].u.memref.shm = shm;
-	submit_command_params[1].u.memref.shm_offs = len;
+	submit_command_params[1].u.memref.size = FTPM_RESPONSE_SIZE_MAX;
+	submit_command_params[1].u.memref.shm = tpm_tee_context.shm;
+	submit_command_params[1].u.memref.shm_offs = tpm_tee_context.response_offset;
 
 	rc = tee_client_invoke_func(
```

Comment:
Is there a guarantee that no two threads can concurrently attempt to send/receive data to/from the TPM, seeing as you removed locking around the send/receive buffer?

Reply:
Nope, that sounds like the right stuff. I figured that there was a mutex higher up the stack, but didn't know for sure.